### PR TITLE
docs: ADR 15 Namespace ID Size

### PR DESCRIPTION
**Pull request by rootulp**
ADR for https://github.com/celestiaorg/celestia-app/issues/1308

### CODE REVIEW

File: docs/architecture/adr-014-namespace-id-size.md
Comment (rootulp) [inline docs/architecture/adr-014-namespace-id-size.md:15]: @musalbas @liamsi do ya'll have any input here?

File: docs/architecture/adr-014-namespace-id-size.md
Comment (musalbas) [inline docs/architecture/adr-014-namespace-id-size.md:9]: Do we want to consider increasing the number of reserved namespace IDs to 65536 or larger, if we also increase namespace ID size?

Comment (rootulp) [reply docs/architecture/adr-014-namespace-id-size.md:9]: Good question. Currently the reserved namespace range occupies:

```go
MinReservedNamespace := namespace.ID{0, 0, 0, 0, 0, 0, 0, 0}
MaxReservedNamespace = namespace.ID{0, 0, 0, 0, 0, 0, 0, 255}
```

and it currently uses 3 of the 256 range for transactions, PFB transactions, and max reserved namespace for reserved namespace padding. Given such low utilization (3/256 is 1% utilization) and without any proposals that increase utilization, I think it's safe to preserve the 256 range and expand at a later date if needed. Thoughts?

For completeness, the protocol also reserves these two:

```
TailPaddingNamespaceID = namespace.ID{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFE}
ParitySharesNamespaceID = namespace.ID{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF}
```

Comment (musalbas) [reply docs/architecture/adr-014-namespace-id-size.md:9]: I think this reserved namespace range is only for namespace version 0 right (assuming this adr is merged)? So if we add extra namespace versions we have to explicitly define the reserved namespace range for those

Comment (rootulp) [reply docs/architecture/adr-014-namespace-id-size.md:9]: Exactly. I expect the strategy you proposed to work if we wanted to increase the reserved namespace range via a namespace id version bump.

Comment (rootulp) [reply docs/architecture/adr-014-namespace-id-size.md]: With notable exceptions for tail padding and parity shares namespaces: https://github.com/celestiaorg/celestia-app/pull/1382#discussion_r1112247526

File: docs/architecture/adr-015-namespace-id-size.md
Comment (evan-forbes) [inline docs/architecture/adr-015-namespace-id-size.md:26]: random kneeeeees!
```suggestion
- [UUIDs](https://en.wikipedia.org/wiki/Universally_unique_identifier) have slightly less than 16 bytes of randomness  and are considered "unique enough for practical purposes" ([Towards Data Science](https://towardsdatascience.com/are-uuids-really-unique-57eb80fc2a87)).
```

**Review by evan-forbes — APPROVED** [review]
nice, makes sense to me! afaiu, nmt, celestia-app, and celestia-node should be able to have larger namespaces with only minor changes here and celestia-node, but I could be wrong.

**Comment by musalbas** [comment]
Should we not agree on the actual new namespace size before merging this or marking as ready for review? The decision section is empty.

**Comment by rootulp** [comment]
> Should we not agree on the actual new namespace size before merging this or marking as ready for review?

I was hoping we could reach a decision via feedback on this PR. If you think we need a meeting to decide, I can set one up.

File: docs/architecture/adr-015-namespace-id-size.md
Comment (staheri14) [inline docs/architecture/adr-015-namespace-id-size.md:24]: > unique namespace IDs (e.g. `sha256(blob)`) with the SHA256 hash function.

It would be helpful to clarify that even in the case of sha256(blob), the namespace ID uniqueness still depends on the blob uniqueness. For example, if two blobs are identical, their hashes will also be identical.

File: docs/architecture/adr-015-namespace-id-size.md
Comment (staheri14) [inline docs/architecture/adr-015-namespace-id-size.md:32]: Can you please add a short description of what column titled `% of 512 byte share` represents.

File: docs/architecture/adr-015-namespace-id-size.md
Comment (staheri14) [inline docs/architecture/adr-015-namespace-id-size.md:51]: Does the current ADR align with the concept of versioned namespace ID? If so, could you please provide more details regarding the last question, which asks about increasing the namespace ID after the mainnet launch? Wouldn't this involve incrementing the version number?

File: docs/architecture/adr-015-namespace-id-size.md
Comment (staheri14) [inline docs/architecture/adr-015-namespace-id-size.md:23]: Can you please provide a link to the `Desirandum` definition? or add a short explanation in the text on what `satisfy desirandum 1` means.

Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:23]: I spelled it incorrectly but I meant to use: https://www.dictionary.com/browse/desideratum
I can replace with something more explicit like "Desirable criteria"

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:32]: Will do! This column describes how much of the share is occupied by the namespace ID. In other words 8/512 or 32/512. The values are represented as a percentage

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:51]: > Does the current ADR align with the concept of versioned namespace ID?

ADR 15 was created after ADR 14 because we aligned on ADR 14 with the assumption that we'll also increase the namespace ID size (hence this ADR). So yes both ADRs are related and aligned.

> If so, could you please provide more details regarding the last question, which asks about increasing the namespace ID after the mainnet launch? Wouldn't this involve incrementing the version number?

Yes I expect it to involve incrementing a version number. Given that, I'll move this question back to ADR 14 and elaborate more there.

Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:51]: > Does the current ADR align with the concept of versioned namespace ID?

ADR 15 was created after ADR 14 because we aligned on ADR 14 with the assumption that we'll also increase the namespace ID size (hence this ADR). So yes both ADRs are related and aligned.

> If so, could you please provide more details regarding the last question, which asks about increasing the namespace ID after the mainnet launch? Wouldn't this involve incrementing the version number?

Yes I expect it to involve incrementing a version number. Given that, I'll move this question back to ADR 14 and elaborate more there.

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md]: Resolved by https://github.com/celestiaorg/celestia-app/pull/1405/commits/1241dd04e85cf6097ed7d5c167cb5e7d8d9f2895

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md]: Resolved by https://github.com/celestiaorg/celestia-app/pull/1405/commits/1241dd04e85cf6097ed7d5c167cb5e7d8d9f2895

File: docs/architecture/adr-015-namespace-id-size.md
Comment (liamsi) [inline docs/architecture/adr-015-namespace-id-size.md:17]:
```suggestion
1. Ability to randomly generate a namespace that hasn't been used before (with large enough probability)
```

Other cirteria:

- Potentially large enough for all rollups that will (ever) exist 
- use a truncated secure hash function (like sha256) for generating the namespace 

The 2nd though is basically the same as "Ability to randomly generate a namespace that hasn't been used before".

Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:17]: > Potentially large enough for all rollups that will (ever) exist

Will add.

> The 2nd though is basically the same as "Ability to randomly generate a namespace that hasn't been used before".

Agreed so there is a bullet point for it in the Notes section but I didn't include it as an explicit desirable criteria.

File: docs/architecture/adr-015-namespace-id-size.md
Comment (SweeXordious) [inline docs/architecture/adr-015-namespace-id-size.md:97]: [Question]
I missed this point, what do you mean by:
 
> wouldn't be sent over the wire

They will be deterministic in some sense?

Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:97]: The optimization would reserve the first N bytes to be unused initially. For example: if the total namespace ID size was 32 bytes, reserve the first 16 bytes as unusable initially. The first 16 bytes would be set to `0`. The last 16 bytes would contain a user's desired namespace.

When the share is sent across the network, a client could receive the first byte (the namespace version) and read that its of namespace version `0` so it can interpret the next 16 bytes as `0` and then read the next `16` bytes as the real namespace. In this optimization, the first 16 bytes of `0` don't need to be sent over the network. Does that make sense?

Comment (SweeXordious) [reply docs/architecture/adr-015-namespace-id-size.md:97]: Defintly, thanks :+1: :+1:

Comment (liamsi) [reply docs/architecture/adr-015-namespace-id-size.md:97]: The 0s also wouldn't be stored, right?

Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:97]: If they aren't stored, then I don't understand where the namespace ID size actually represents 32 bytes. With this optimization, I expect the leading 16 bytes of `0`s to be stored in the share. 

Can you clarify where you don't think they would be stored @liamsi ?

Comment (musalbas) [reply docs/architecture/adr-015-namespace-id-size.md:97]: I don't think they need to be stored on disk, for example.

The only place where the 0s would be recognized is when computing the hash of intermediate nodes.

That being said, for this to solve the backwards-compatibility issue we're talking about, rollup software and NMT implementations would have to not assume that all intermediate nodes will have 0s at the start, in case we do ever increase the namespace size as use those preceding 0s...

Comment (musalbas) [reply docs/architecture/adr-015-namespace-id-size.md:97]: For example, assuming hashes are big endian, then when serializing the hashes over the wire or on disk, preceding 0s can be omitted, just like how Ethereum addresses that start up with 0s are more gas efficient

Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:97]: > The only place where the 0s would be recognized is when computing the hash of intermediate nodes.

I think the 0s (or at least the share version) would also need to be recognized when creating a share commitment so that these two shares don't end up resulting in the same share commitment:

```go
shareA := []byte {
	0x0 // share version
	0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, // 16 bytes of unrecognized namespace
	0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, // 16 bytes of user specified namespace
	// remainder of share ommited for brevity
}
shareB := []byte {
	0x1 // share version
	0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, // 16 bytes of user specified namespace
	0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, // another 16 bytes of user specified namespace
	// remainder of share ommited for brevity
}
```

if that's the case, we're special casing the share version `0` to be interpreted as 16 bytes of `0`s in most instances where the namespace ID is used. In order to evaluate this optimization, I think we need to consider the code complexity and investigate how much bandwidth this decreases (presumably 16 bytes every time a namespace ID is sent over the network) but idk how much that is relative to total bandwidth requirements.

Comment (musalbas) [reply docs/architecture/adr-015-namespace-id-size.md:97]: Oh I see, yeah you can also make the 16 0s enforced by the share version, and if we increase the namespace size, introduce a new share version (which is opt-in by rollups thus it doesn't break backwards compatibility)

That being said, if a Merkle proof in an old share version contains a sibling for a new share version that has a longer namespace, that could break backwards compatibility, so you may need a bunch of padding shares ... (or I think the non-interactive defaults already solve this issue ?)

Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:97]: > That being said, if a Merkle proof in an old share version contains a sibling for a new share version that has a longer namespace, that could break backwards compatibility

Agreed this does break backwards compatability. I don't expect non-interactive defaults to solve the issue. An NMT proof doesn't appear to contain the raw share of a sibling based on [these docs](https://github.com/celestiaorg/nmt/blob/1bc0bb0099e01b30e37ddb56642734ae875917cd/nmt.go#L150-L155) but `proof.nodes` contains namespaced hashes so if shareA and shareB are in the same data square and correspondingly same NMT, then inner nodes of the NMT for `namespaceVersion=0`  would be of size: 2 * 16 + 32 = 64 and inner nodes of the NMT for `namespaceVersion=1` would be of size: 2 * 32 + 32 = 96.

Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:97]: In order for NMT proof verification to work, the original NMT proof verification logic would need to interpret `namespaceVersion=0` with 16 leading bytes of `0`s and to not perform such interpretation for `namespaceVersion=1` and above. It seems contrived to implement NMT proof verification in this way and it may in fact break if we made a different change when introducing `namespaceVersion=1`.

Preserving backwards compatibility of proofs for an NMT constructed with different size namespaces seems brittle in this scenario. Thoughts? @musalbas

Comment (musalbas) [reply docs/architecture/adr-015-namespace-id-size.md:124]: Yeah, you wouldn't have backwards compatibility if you have different namespace sizes for different versions anyways, because at least the root node might have a max or min node of a different size than older clients expect.

I think the only backwards compatible way is if all clients don't make any assumption about how many leading 0s there are, but have a consensus rule that requires leading 0s, like the Ethereum address optimization I mentioned here

>For example, assuming hashes are big endian, then when serializing the hashes over the wire or on disk, preceding 0s can be omitted, just like how Ethereum addresses that start with 0s are more gas efficient

**Comment by rootulp** [comment]
CI markdownlint doesn't like the footnotes: https://github.com/celestiaorg/celestia-app/actions/runs/4318011765/jobs/7535787254#step:4:32 but local markdownlint doesn't complain

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:172]: Correct. The namespace ID size in the NMT will be 33 bytes. 

Users will specify a namespace version (1 byte / uint8) and a namespace ID (32 bytes) in their PFB. Additionally we should strive to make it clear to users that these two will be interpreted as distinct namespace IDs:

```go
namespaceVersion := 0
namespaceId := []byte{1, 1, 1, ...} // 32 bytes

namespaceVersion := 1
namespaceId := []byte{1, 1, 1, ...} // 32 bytes
```

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:168]: Can do.

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md]: Clarified in https://github.com/celestiaorg/celestia-app/pull/1405/commits/4a91bf6ace87bd91820be2b7794d9276ae2c9eea

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md]: Done in https://github.com/celestiaorg/celestia-app/pull/1405/commits/4a91bf6ace87bd91820be2b7794d9276ae2c9eea

**Review by evan-forbes — APPROVED** [review]

**Comment by rootulp** [comment]
Will merge by EOD if there is no more feedback

File: docs/architecture/adr-015-namespace-id-size.md
Comment (staheri14) [inline docs/architecture/adr-015-namespace-id-size.md:131]: [Sharing thoughts] The total size of a blob is also affected by the ID size. For example, if a 32-byte ID is used instead of an 8-byte ID, the total size of a blob is reduced by 128 * (32-8) (128 refers to the max number of shares per blob as mentioned earlier).

File: docs/architecture/adr-015-namespace-id-size.md
Comment (staheri14) [inline docs/architecture/adr-015-namespace-id-size.md:145]: [Question] I'm uncertain how we could incorporate this idea into NMT, since the block shares' corresponding leaves in NMT must have distinct namespaces. Consequently, we may not be able to apply this optimization in the final NMT representation.

File: docs/architecture/adr-015-namespace-id-size.md
Comment (staheri14) [inline docs/architecture/adr-015-namespace-id-size.md:148]:
```suggestion
1. What are the performance implications on celestia-node for a larger namespace ID size?
2. Is it possible to mitigate some tradeoffs when adopting a large namespace ID size?
    1. It may be possible to decrease the bandwidth requirements for NMT proofs by using lossless compression (proposed by @evan-forbes) and explored above.
    2. It may be possible to avoid writing the namespace ID to continuation blob shares (proposed by @nashqueue)
        1. Note this introduces complexity for erasure reconstruction. A share in row B may have its namespace in row A so to reconstruct a data square, we must refactor the process to two steps:
            1. Reconstruct all shares from the erasure coding
            2. Reconstruct the NMT
```

File: docs/architecture/adr-015-namespace-id-size.md
Comment (staheri14) [inline docs/architecture/adr-015-namespace-id-size.md:143]: [nit] All the items have been assigned the same number "1", it needs to be fixed.

File: docs/architecture/adr-015-namespace-id-size.md
Comment (staheri14) [inline docs/architecture/adr-015-namespace-id-size.md:22]: [Question] Can we have shares with different namespace versions in the same NMT?

**Review by staheri14 — APPROVED** [review]
Nice! left some non-blocking comments.

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:26]: resolved by cc14b43c87fe2b8069272483ba6e394ada6793bb

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:28]: Yes. Such a rule would be consensus-critical (likely enforced in process_proposal.go).

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:36]: these two refers to the two namespace IDs below. I'll re-word the section below

Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md]: resolved by 3cbd49eea640d20d28e052ddaf6b79f5bfb4f384

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:51]: [informational] Markdown lets one use the `1.` prefix for multiple entires in an [ordered list](https://www.markdownguide.org/basic-syntax/#ordered-lists). I prefer the `1.` prefix because if you re-order items in the list, you don't need to manually change each number. Will accept suggestion so clearer for readers.

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md]: resolved by 5e136bac34e044d799bc56e7025b27a2a8239213

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:120]: thanks, will add a note below this section.

Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:120]: resolved by f94c7f249e4a592e46622c3bad84c41ef2f3961b

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:149]: Yea, this optimization would happen before writing to the leaf. There's a Slack thread exploring this topic here: https://celestia-team.slack.com/archives/C02PFFCGMNW/p1677197644501209

The rough idea is that shares wouldn't contain the namespace ID but the entire data square (or at least preceding rows) would need to be reconstructed prior to constructing the NMT so that leaves pushed to the NMT would contain the namespace ID.

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md]: resolving b/c https://www.markdownguide.org/basic-syntax/#ordered-lists

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:22]: Depends on what change the new namespace version is introduced for but if a new namespace version is introduced to change something like non-interactive default rules then yes b/c both namespace version `0` and `1` will be of size 32 bytes (ID) + 1 byte (version) = 33 bytes.

File: docs/architecture/adr-015-namespace-id-size.md
Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:135]: Yea good point. Will add a section for it.

Comment (rootulp) [reply docs/architecture/adr-015-namespace-id-size.md:135]: Resolved by 70fb1308288b7bcb2a4bfbe152496b16aea89a98

**Review by evan-forbes — APPROVED** [review]

**Review by staheri14 — APPROVED** [review]
Thanks for addressing the comments, LGTM!

**Pull request merged**
